src/rctl.rs

```diff
@@ -0,0 +1,1595 @@
+    /// to be compiled without the `RCTL` / `RACCT` options.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use jail::rctl;
+    /// let state = rctl::State::check();
+    /// ```
+    pub fn check() -> State {
+        let enable_racct = sysctl::Ctl::new("kern.racct.enable");
+
+        // If the sysctl call fails, we assume it to be disabled.
+        if enable_racct.is_err() {
+            return State::NotPresent;
+        }
+
+        match enable_racct.unwrap().value() {
+            Ok(sysctl::CtlValue::Uint(v)) => match v {
+                1 => State::Enabled,
+                _ => State::Disabled,
+            },
+            // If we could not get the sysctl value, assume it to be disabled.
+            _ => State::NotPresent,
+        }
+    }
+
+    /// Return `true` if the `RCTL` / `RACCT` support is [Enabled].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use jail::rctl;
+    /// if rctl::State::check().is_enabled() {
+    ///     // do things requiring `RCTL` / `RACCT` support.
+    /// }
+    /// ```
+    ///
+    /// [Enabled]: State::Enabled
+    pub fn is_enabled(&self) -> bool {
+        match self {
+            State::Enabled => true,
+            _ => false,
+        }
+    }
+
+    /// Return `true` if the kernel has `RCTL` / `RACCT` support compiled in.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use jail::rctl;
+    /// if ! rctl::State::check().is_present() {
+    ///     println!("The kernel does not have RCTL / RACCT support");
+    /// }
+    /// ```
+    pub fn is_present(&self) -> bool {
+        match self {
+            State::NotPresent => false,
+            _ => true,
+        }
+    }
+}
+
+/// Get the resource usage for a specific [Subject].
+///
+/// # Example
+///
+/// ```
+/// # extern crate jail;
+/// # use jail::rctl;
+/// extern crate libc;
+///
+/// let uid = unsafe { libc::getuid() };
+/// let subject = rctl::Subject::user_id(uid);
+///
+/// let usage = rctl::usage(subject)
+///     .expect("Could not get RCTL usage");
+///
+/// println!("{:#?}", usage);
+/// ```
+pub fn usage(subject: Subject) -> Result<HashMap<Resource, usize>, Error> {
+    extern "C" {
+        fn rctl_get_racct(
+            inbufp: *const libc::c_char,
+            inbuflen: libc::size_t,
+            outbufp: *mut libc::c_char,
+            outbuflen: libc::size_t,
+        ) -> libc::c_int;
+    }
+
+    let filter = Filter::new().subject(subject).to_string();
+    let filterlen = filter.len() + 1;
+
+    // Get the filter as a C string.
+    let cfilter = CString::new(filter).map_err(Error::CStringError)?;
+
+    // C compatible output buffer.
+    let outbuflen: usize = RCTL_DEFAULT_BUFSIZE / 4;
+    let mut outbuf: Vec<i8> = vec![0; outbuflen];
+
+    // Unsafe C call to get the jail resource usage.
+    if unsafe { rctl_get_racct(cfilter.as_ptr(), filterlen, outbuf.as_mut_ptr(), outbuflen) } != 0 {
```

Comment:
from [`ractl_get_racct`(2)]:
> **RULES AND FILTERS**
> This section explains how the rule and filter format described in rctl(8) differs from the format passed to the system calls themselves.  The rctl tool provides several conveniences that the system calls do not.  When  using the system call:
>
>   *   The subject must	be fully specified.  For example, abbreviating  `user` to `u` is	not acceptable.
>   *   User and	group IDs must be numeric.  For	example, `root`	must be expressed as `0`.
>   *  Units are not permitted on resource amounts.  For example, a quantity	of 1024 bytes must be expressed	as `1024` and not `1k`.

The `fmt::Display` implementation here prints these things in the humanized form - this works well for jails, but won't work for users.

[`ractl_get_racct`(2)]: https://www.freebsd.org/cgi/man.cgi?query=rctl_get_racct&sektion=2&manpath=FreeBSD+11.2-RELEASE+and+Ports